src/Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -509,67 +517,82 @@ public void StopStake()
                 }
                 else
                 {
-                    this.logger.LogTrace("{0} failed, waiting {1} ms for next round...", nameof(this.StakeAndSignBlockAsync), this.minerSleep);
-                    await Task.Delay(TimeSpan.FromMilliseconds(this.minerSleep), this.stakeCancellationTokenSource.Token).ConfigureAwait(false);
+                    this.logger.LogTrace("{0} failed to create POS block, waiting {1} ms for next round...", nameof(this.StakeAndSignBlockAsync), this.minerSleep);
+                    await Task.Delay(TimeSpan.FromMilliseconds(this.minerSleep), cancellationToken).ConfigureAwait(false);
                 }
             }
         }
 
+        public async Task<List<UtxoStakeDescription>> GetUtxoStakeDescriptionsAsync(WalletSecret walletSecret, CancellationToken cancellationToken)
+        {
+            this.logger.LogTrace("()");
+            var utxoStakeDescriptions = new List<UtxoStakeDescription>();
+            List<UnspentOutputReference> spendableTransactions = this.walletManager
+                .GetSpendableTransactionsInWallet(walletSecret.WalletName, 1).ToList();
+
+            FetchCoinsResponse fetchedCoinSet = await this.coinView.FetchCoinsAsync(spendableTransactions.Select(t => t.Transaction.Id).ToArray(), cancellationToken).ConfigureAwait(false);
+            if (cancellationToken.IsCancellationRequested)
+                return utxoStakeDescriptions;
+
+            foreach (UnspentOutputReference outputReference in spendableTransactions)
+            {
+                UnspentOutputs coinSet = fetchedCoinSet.UnspentOutputs
+                    .FirstOrDefault(f => f?.TransactionId == outputReference.Transaction.Id);
+                if ((coinSet == null) || (outputReference.Transaction.Index >= coinSet.Outputs.Length))
+                    continue;
+
+                TxOut utxo = coinSet.Outputs[outputReference.Transaction.Index];
+                if ((utxo == null) || (utxo.Value <= MinimumStakingCoinValue))
+                    continue;
+
+                uint256 hashBlock = this.chain.GetBlock((int)coinSet.Height)?.HashBlock;
+                if (hashBlock == null)
+                    continue;
+
+                var utxoStakeDescription = new UtxoStakeDescription
+                {
+                    TxOut = utxo,
+                    OutPoint = new OutPoint(coinSet.TransactionId, outputReference.Transaction.Index),
+                    Address = outputReference.Address,
+                    HashBlock = hashBlock,
+                    UtxoSet = coinSet,
+                    Secret = walletSecret // Temporary.
+                };
+                utxoStakeDescriptions.Add(utxoStakeDescription);
+
+                this.logger.LogTrace("UTXO '{0}' with value {1} might be available for staking.", utxoStakeDescription.OutPoint, utxo.Value);
+
+                if (cancellationToken.IsCancellationRequested)
+                    return utxoStakeDescriptions;
```

Comment:
don't return just break and move to the top of foreach

Reply:
OK I don't understand; if you know that the staking loop has been cancelled, why bother going on any further with it ? (that is assuming that by "break" you meant "continue")

Reply:
i mean break, not continue

Reply:
actually, even break is too weak probably as it would just not prevent the caller to continue, therefore i'd suggest
1) move it to the top of the loop
2) make it throw with exit log:
```
if (cancellationToken.IsCancellationRequested)
{
    throw new OperationCancelledException();
    this.logger.LogTrace("(-)[CANCELLATION]");
}
```

note that i'm intentionally (cause of log) not using throw if cancelled on the token